In [1]:
from bs4 import BeautifulSoup as bs
import requests
import pandas as pd
import numpy as np

# Load the Webpage

\
following video solution.

In [2]:
# Load the webpage that contains the tables of all the disney movies
r=requests.get('https://en.wikipedia.org/wiki/List_of_Walt_Disney_Pictures_films')

# Convert to a beautiful Soup object

soup=bs(r.content)

# Print out the HTML content

# contents=soup.prettify()
# print(contents)


# IDEA: Be thinking about what you want to do and use Google to your advantage.

In [3]:
# In the html for the List of Walt Disney Pictures films it looks like the movie titles are hyperlinked
# each title is containted in a table with class="wikitable sortable jquery-tablesorter"
# we can use the .select() method to point to the table.
# Inside .select() '.' can be used to point to a class and the name. 
# If the class has a space (' ') then we can still use '.' to fill in the class name

# each title and link is inside <i> element and <a> so we will point to <a> to get the title text, and links.

movie_list = soup.select('table.wikitable.sortable i a')
movie_list

[<a href="/wiki/Academy_Award_Review_of_Walt_Disney_Cartoons" title="Academy Award Review of Walt Disney Cartoons">Academy Award Review of Walt Disney Cartoons</a>,
 <a href="/wiki/Snow_White_and_the_Seven_Dwarfs_(1937_film)" title="Snow White and the Seven Dwarfs (1937 film)">Snow White and the Seven Dwarfs</a>,
 <a href="/wiki/Pinocchio_(1940_film)" title="Pinocchio (1940 film)">Pinocchio</a>,
 <a href="/wiki/Fantasia_(1940_film)" title="Fantasia (1940 film)">Fantasia</a>,
 <a href="/wiki/The_Reluctant_Dragon_(1941_film)" title="The Reluctant Dragon (1941 film)">The Reluctant Dragon</a>,
 <a href="/wiki/Dumbo" title="Dumbo">Dumbo</a>,
 <a href="/wiki/Bambi" title="Bambi">Bambi</a>,
 <a href="/wiki/Saludos_Amigos" title="Saludos Amigos">Saludos Amigos</a>,
 <a href="/wiki/Victory_Through_Air_Power_(film)" title="Victory Through Air Power (film)">Victory Through Air Power</a>,
 <a href="/wiki/The_Three_Caballeros" title="The Three Caballeros">The Three Caballeros</a>,
 <a href="/wiki/M

In [4]:
# We can get all of the wiki links, and titles by going into movie_list recurrsively (multiple times) using a list comprehension or for loop.

# Look at the first element of of movie_list to get an idea of how to grab the reference (wiki) links.

movie_list[0], movie_list[0]['href'], movie_list[0]['title']

(<a href="/wiki/Academy_Award_Review_of_Walt_Disney_Cartoons" title="Academy Award Review of Walt Disney Cartoons">Academy Award Review of Walt Disney Cartoons</a>,
 '/wiki/Academy_Award_Review_of_Walt_Disney_Cartoons',
 'Academy Award Review of Walt Disney Cartoons')

In [5]:
# Make a list comprehension that grabs all of the titles and links and stores the pair labels as dictionaries.

title_href_list = [{'title':movie['title'], 'href':movie['href']} for movie in movie_list]
title_href_list

[{'title': 'Academy Award Review of Walt Disney Cartoons',
  'href': '/wiki/Academy_Award_Review_of_Walt_Disney_Cartoons'},
 {'title': 'Snow White and the Seven Dwarfs (1937 film)',
  'href': '/wiki/Snow_White_and_the_Seven_Dwarfs_(1937_film)'},
 {'title': 'Pinocchio (1940 film)', 'href': '/wiki/Pinocchio_(1940_film)'},
 {'title': 'Fantasia (1940 film)', 'href': '/wiki/Fantasia_(1940_film)'},
 {'title': 'The Reluctant Dragon (1941 film)',
  'href': '/wiki/The_Reluctant_Dragon_(1941_film)'},
 {'title': 'Dumbo', 'href': '/wiki/Dumbo'},
 {'title': 'Bambi', 'href': '/wiki/Bambi'},
 {'title': 'Saludos Amigos', 'href': '/wiki/Saludos_Amigos'},
 {'title': 'Victory Through Air Power (film)',
  'href': '/wiki/Victory_Through_Air_Power_(film)'},
 {'title': 'The Three Caballeros', 'href': '/wiki/The_Three_Caballeros'},
 {'title': 'Make Mine Music', 'href': '/wiki/Make_Mine_Music'},
 {'title': 'Song of the South', 'href': '/wiki/Song_of_the_South'},
 {'title': 'Fun and Fancy Free', 'href': '/wiki/

In [6]:
# Let's look at how many movies, and references there are.
count_movies = len(title_href_list)
count_movies

449

## Function: get_soup(url)

diverting from video solution.

In [7]:
# make a new method make_soup(url) that will go to the url and create a beautiful soup object.
def make_soup(url):
    r=requests.get(url)
    soup=bs(r.content)
    
    return soup
    

In [8]:
# Let's try this method with the first movie in title_href_list
# Since the hrefs are only the relative path for the movie pages, we will need to contruct the full path for the url variable.

relative_path = title_href_list[0]['href']
base_path = 'https://en.wikipedia.org/'
url = base_path + relative_path
url

'https://en.wikipedia.org//wiki/Academy_Award_Review_of_Walt_Disney_Cartoons'

In [9]:
soup=make_soup(url)

In [10]:
# from the soup we will want to find the html that contains the info_box information
# from inspect it looks like the info_box is stored as class="infobox vevent"
# all of the table rows contain the label, and data we need

soup.select('table.infobox.vevent tr')


[<tr><th class="summary" colspan="2" style="text-align:center;font-size:125%;font-weight:bold;font-size:110%;font-style:italic;">Academy Award Review of <span class="nowrap">Walt Disney Cartoons</span></th></tr>,
 <tr><td colspan="2" style="text-align:center"><a class="image" href="/wiki/File:Aarfvdsf.jpg"><img alt="Aarfvdsf.jpg" class="thumbborder" data-file-height="111" data-file-width="300" decoding="async" height="81" src="//upload.wikimedia.org/wikipedia/en/thumb/7/7a/Aarfvdsf.jpg/220px-Aarfvdsf.jpg" srcset="//upload.wikimedia.org/wikipedia/en/7/7a/Aarfvdsf.jpg 1.5x" width="220"/></a><div style="font-size:95%;padding:0.35em 0.35em 0.25em;line-height:1.25em;">Opening and closing screens</div></td></tr>,
 <tr><th scope="row" style="white-space:nowrap;padding-right:0.65em;"><div style="display:inline-block; padding:0.1em 0;line-height:1.2em;">Production<br/>company </div></th><td><div style="vertical-align:middle;"><a href="/wiki/Walt_Disney_Animation_Studios" title="Walt Disney Anim

## Function: make_wiki_url(relative_path)


In [11]:
def make_wiki_url(relative_path, base_path = 'https://en.wikipedia.org/'):
    url= base_path + relative_path
    return url

In [12]:
movie_wiki_link_list = [make_wiki_url(movie['href']) for movie in title_href_list]
movie_wiki_link_list

['https://en.wikipedia.org//wiki/Academy_Award_Review_of_Walt_Disney_Cartoons',
 'https://en.wikipedia.org//wiki/Snow_White_and_the_Seven_Dwarfs_(1937_film)',
 'https://en.wikipedia.org//wiki/Pinocchio_(1940_film)',
 'https://en.wikipedia.org//wiki/Fantasia_(1940_film)',
 'https://en.wikipedia.org//wiki/The_Reluctant_Dragon_(1941_film)',
 'https://en.wikipedia.org//wiki/Dumbo',
 'https://en.wikipedia.org//wiki/Bambi',
 'https://en.wikipedia.org//wiki/Saludos_Amigos',
 'https://en.wikipedia.org//wiki/Victory_Through_Air_Power_(film)',
 'https://en.wikipedia.org//wiki/The_Three_Caballeros',
 'https://en.wikipedia.org//wiki/Make_Mine_Music',
 'https://en.wikipedia.org//wiki/Song_of_the_South',
 'https://en.wikipedia.org//wiki/Fun_and_Fancy_Free',
 'https://en.wikipedia.org//wiki/Melody_Time',
 'https://en.wikipedia.org//wiki/So_Dear_to_My_Heart',
 'https://en.wikipedia.org//wiki/The_Adventures_of_Ichabod_and_Mr._Toad',
 'https://en.wikipedia.org//wiki/Cinderella_(1950_film)',
 'https://en

## Function: make_table_rows(soup)


In [13]:
def make_table_rows(soup):
    table_rows = soup.select('table.infobox.vevent tr')
    return table_rows

## Function: get_info_boxes(href_list)

In [14]:
# create a function that takes in a list of relative wiki path url's.
# create wiki url paths with make_wiki_url(relative_path, base_path = "https://en.wikipedia.org/")
# create soup objecs using make_soup(url) for each link in make_wiki_link_list
# create a list of info_boxes using make_table_rows(soup) for each soup in soups (list).
# return back the list of info_boxes
def get_info_boxes(href_list):
    
    movie_wiki_link_list=[make_wiki_url(movie['href']) for movie in href_list]
    
    soups = [make_soup(url) for url in movie_wiki_link_list]
    
    info_boxes = [make_table_rows(soup) for soup in soups]
    
    return info_boxes
    

In [15]:
# create all the info boxes
info_boxes = get_info_boxes(title_href_list)

In [16]:
# [clean]
# get rid of subscripts ([1],[2],etc.)

def remove_tags(content):
    # [clean] removing subscripts:
    if content.find_all(['sup','span']):
        for tag in content.find_all('sup'):
            tag.decompose()
                        
    content = content.get_text(" ", strip=True).replace('\xa0', ' ')
    
    return content

In [17]:
[(index, movie) for index, movie in enumerate(title_href_list) if movie['title'] == 'Sacred Planet']

[(276, {'title': 'Sacred Planet', 'href': '/wiki/Sacred_Planet'})]

In [18]:
# Testing
[remove_tags(row) for row in info_boxes[276]]

['Sacred Planet',
 'DVD',
 'Directed by Jon Long Hairul Salleh Askor',
 'Produced by Jon Long Karen Fernandez Long',
 'Written by Karen Fernandez Long Jon Long',
 'Starring Arapata McKay Tsaan Ciqae Mae Tui Cy Peck Jr. Mutang Urud',
 'Narrated by Robert Redford',
 'Cinematography William Reeve',
 'Edited by Jon Long',
 'Production company Walt Disney Pictures',
 'Distributed by Buena Vista Pictures',
 'Release date April 22, 2004 ( 2004-04-22 )',
 'Running time 40 minutes',
 'Country Canada Malaysia United States',
 'Language English',
 'Box office $1,108,356']

## Function: get_br_list(td), get_content_value(content)

In [19]:
def get_br_list(td):

    if len(td.find_all('br')) > 1: 
        content_value = ([str(td.find_all('br')[x].previousSibling) for x in range(len(td.find_all('br')))])
        content_value.append(str(td.find_all('br')[-1].next_sibling))
    elif len(td.find_all('br')) == 1 :
        a = []
        [(a.append(str(td.find_all('br')[x].previousSibling)), a.append(str(td.find_all('br')[x].nextSibling))) for x in range(len(td.find_all('br')))]
        content_value=a
    
    return [x.lower() for x in content_value]

In [20]:
def get_content_value(content):
    # find the <td> element in the bs element
    content_value = content.find('td')

    # if the <td> element contains list elements <li>: then put the text from those elements into a list named content_value
    # remove the subscripts from any list elements using remove_tags(content)
    if content_value.find('li'):
        li_list = content_value.find_all('li')
                    
        # [clean] subscripts out of li items
        content_value = [remove_tags(x).lower() for x in li_list]
    
    # if there are <br> elements and have links <a href=>: store them in a list named content_value.
    # else: store the value in a variable named content_value
    elif content_value.find('br'):
        # ['clean'] get long strings into a list. 
        # These items were not storring as lists such as in the 'Starting' key
        
        if content_value.find('a'):     
            a_list = content_value.find_all('a')
            content_value= [remove_tags(x).lower() for x in a_list]
        else:
            content_value= get_br_list(content_value)

#             content_value = remove_tags(content_value)
    # Else: get the text from the table td.                                
    else:
        content_value=remove_tags(content_value).lower()
        
    return content_value

In [21]:
test = 'apple'
test.lower()

'apple'

In [22]:
# testing getting text for 'td' elements that contain 'br'
# info_boxes[276]

In [23]:
td =info_boxes[276][2].find('td')
td

<td>Jon Long<br/>Hairul Salleh Askor</td>

In [24]:
get_br_list(td)

['jon long', 'hairul salleh askor']

In [25]:
# test
[print(index, get_content_value(row)) for index, row in enumerate(info_boxes[276]) if index > 1]

2 ['jon long', 'hairul salleh askor']
3 ['jon long', 'karen fernandez long']
4 ['karen fernandez long', 'jon long']
5 ['arapata mckay', 'tsaan ciqae', 'mae tui', 'cy peck jr.', 'mutang urud']
6 robert redford
7 william reeve
8 jon long
9 walt disney pictures
10 buena vista pictures
11 ['april 22, 2004 ( 2004-04-22 )']
12 40 minutes
13 ['canada', 'malaysia', 'united states']
14 english
15 $1,108,356


[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None]

In [26]:
# Create a dictionary that contains the keys and values of movie information
def create_movie_dictionary(row_data):
    movie_info={}
    for index, row in enumerate(row_data):
        if index==0:
            movie_info['title'] = row.find("th").get_text().lower()
        elif index > 1:
            try:
                content_key = row.find('th').get_text(" ", strip=True)

                content_value = get_content_value(row)

                movie_info.update({content_key:content_value})

            except Exception as e:
                continue
        
    return movie_info

In [27]:
create_movie_dictionary(info_boxes[14])

{'title': 'so dear to my heart',
 'Directed by': ['harold d. schuster', 'hamilton luske'],
 'Produced by': ['walt disney', 'perce pearce'],
 'Written by': ['ken anderson',
  'marc davis',
  'bill peet',
  'maurice rapf',
  'ted sears'],
 'Based on': ['midnight and jeremiah', 'sterling north'],
 'Starring': ['bobby driscoll', 'luana patten', 'beulah bondi', 'burl ives'],
 'Music by': 'paul smith',
 'Cinematography': 'winton c. hoch',
 'Edited by': ['lloyd l. richardson', 'thomas scott'],
 'Production company': 'walt disney productions',
 'Distributed by': 'rko radio pictures, inc.',
 'Release date': ['[1]'],
 'Running time': '82 minutes',
 'Country': 'united states',
 'Language': 'english',
 'Budget': '$1.5 million',
 'Box office': '$3.7 million (u.s. rental) $575,000 (foreign rental)'}

In [28]:
# create a movie dictionary for each movie in info_boxes. Store the dictionaries in list named 'df' to represent the data frame.
df = [create_movie_dictionary(movie) for movie in info_boxes]

In [29]:
[(index,x) for index,x in enumerate(df) if 'title' not in x]

[(124, {}), (440, {})]

In [30]:
# the length of df (count of dictionaries) matches the title_href_list length (count movies)
len(df)

449

In [31]:
# check to see if 'title' appears in all the movie dictionaries
[(index,movie) for index, movie in enumerate(df) if 'title' not in movie]

[(124, {}), (440, {})]

In [32]:
# what are the href of these two movies?
movie_list[124], movie_list[440]

(<a href="/wiki/True-Life_Adventures" title="True-Life Adventures">True-Life Adventures</a>,
 <a class="mw-redirect" href="/wiki/Luca_(2021_film)" title="Luca (2021 film)">Luca</a>)

In [33]:
# these two pages are not movie pages they are a series page (True-Life Adventures), and Pixar Page (Luca).
# We will drop them from the df.
info_boxes[124], info_boxes[440]

([], [])

In [34]:
# renaming df to only be dictionaries that contain the the title key
df = [movie for movie in df if 'title' in movie]
len(df)

447

In [35]:
# [(index, x['title'], {k:[type(x) for x in v if type(x)!=str] for k,v in x.items()}) for index, x in enumerate(df)]

In [36]:
# How many movies contain do not contain 'Box office' key? This may be a set to drop when we want to analyze later.
len([(index, movie) for index, movie in enumerate(df) if 'Box office' not in movie])

78

## Function: write to json(obj, file_name), load_data(file_name)

In [37]:
# Save list of Dictionaries as a json file.
import json

def write_to_json(obj, file_name):
    with open(file_name, 'w') as fout:
        json.dump(obj, fout, ensure_ascii=False, indent=2)

def load_data(file_name):
    with open(file_name, encoding='utf-8') as f:
        return json.load(f)

In [38]:
# write df to a .json file.
write_to_json(df, 'movie_dictionaries.json')

In [39]:
# load 'movie_dictionaries.json' and store as a variable named df. 
# If 'movie_dictionaries.json' does not exist in the directory this function will fail.
df = load_data('movie_dictionaries.json')

In [40]:
[x for x in df if x['title']=='sacred planet']

[{'title': 'sacred planet',
  'Directed by': ['jon long', 'hairul salleh askor'],
  'Produced by': ['jon long', 'karen fernandez long'],
  'Written by': ['karen fernandez long', 'jon long'],
  'Starring': ['arapata mckay',
   'tsaan ciqae',
   'mae tui',
   'cy peck jr.',
   'mutang urud'],
  'Narrated by': 'robert redford',
  'Cinematography': 'william reeve',
  'Edited by': 'jon long',
  'Production company': 'walt disney pictures',
  'Distributed by': 'buena vista pictures',
  'Release date': ['april 22, 2004 ( 2004-04-22 )'],
  'Running time': '40 minutes',
  'Country': ['canada', 'malaysia', 'united states'],
  'Language': 'english',
  'Box office': '$1,108,356'}]

In [41]:
# check the length of the dataframe loaded.
len(df)

447

In [42]:
#[clean] Testing to see if we are getting 'clean' versions of dictionaries.
[(index, x) for index,x in enumerate(df) if x['title']=="The Vanishing Prairie"]

[]

In [43]:
# [clean] Testing to  see if we are getting 'clean' versions of dictionaries.
df[0:5]

[{'title': 'academy award review of walt disney cartoons',
  'Production company': 'walt disney productions',
  'Release date': ['may 19, 1937 ( 1937-05-19 )'],
  'Running time': '41 minutes (74 minutes 1966 release)',
  'Country': 'united states',
  'Language': 'english',
  'Box office': '$45.472'},
 {'title': 'snow white and the seven dwarfs',
  'Directed by': ['david hand (supervising)',
   'william cottrell',
   'wilfred jackson',
   'larry morey',
   'perce pearce',
   'ben sharpsteen'],
  'Produced by': 'walt disney',
  'Written by': ['ted sears',
   'richard creedon',
   'otto englander',
   'dick rickard',
   'earl hurd',
   'merrill de maris',
   'dorothy ann blank',
   'webb smith'],
  'Based on': ['snow white', 'brothers grimm'],
  'Starring': ['adriana caselotti',
   'lucille la verne',
   'harry stockwell',
   'roy atwell',
   'pinto colvig',
   'otis harlan',
   'scotty mattraw',
   'billy gilbert',
   'eddie collins',
   'moroni olsen',
   'stuart buchanan'],
  'Music by

# Clean the Data!

- Standarize subscripts, Clean up references
- Standarize Time 
    - Convert running time to an integer
    - convert data as date-time object
- Standarize Money
    - Convert budget
    - box office to numbers
- Standarize 'Starring' make a list

In [44]:
# Clean up references. ([1] [2] etc.)
# Going to clean up directly in create_movie_dictionary(row_data) above.
# Will make comments about cleaning as [clean].
# Done

In [45]:
# Split up the long strings where multiple names are not put into lists such as in the 'Starring' key
# Will make comments about cleaning as [clean].


In [46]:
# Standardize Time
# Convert running time to an integer.
# Let's see how 'One Little Indian' turned out.
little_indian = [x for x in df if x['title']=="one little indian"]
little_indian

[{'title': 'one little indian',
  'Directed by': 'bernard mceveety',
  'Produced by': 'winston hibler',
  'Written by': 'harry spalding',
  'Starring': ['james garner',
   'vera miles',
   'pat hingle',
   'morgan woodward',
   'jodie foster'],
  'Music by': 'jerry goldsmith',
  'Cinematography': 'charles f. wheeler',
  'Edited by': 'robert stafford',
  'Production company': 'walt disney productions',
  'Distributed by': 'buena vista distribution',
  'Release date': ['june 20, 1973 ( 1973-06-20 )'],
  'Running time': '90 minutes',
  'Country': 'united states',
  'Language': 'english',
  'Box office': '$2 million'}]

In [47]:
# Navigate into the Running time value of 'One Little Indian'
running_time = little_indian[0]['Running time']
running_time

'90 minutes'

In [48]:
# How can we clean up the 'Running time' value to be an integer?
# Replace space (' ') and 'Minutes' with no values. 
# wrap the code with int() to change the type from string to an integer.
int(running_time.replace(' ', "").replace('minutes', ""))

90

In [49]:
# How many movies have the 'Running time' field
movies_with_runtime = [x for x in df if 'Running time' in x]
len(movies_with_runtime)

434

In [50]:
# strip running time if it is a string at '' 'min' (include the space so you don't have to strip it later). 
# movies_with_runtime = [{'title': x['title'], 'Running time': x['Running time'].split(' min', 1)[0]} if type(x['Running time']) is str else {'title':x['title']} for x in  movies_with_runtime ]

In [51]:
# remove the movies that don't have the runtime as a string:
# NOTE: '-' is a HYPHEN will take copy the output character to use in the split below.
# [{'title': x['title'], 'Running time': x['Running time'].split(' min', 1)[0]} for x in movies_with_runtime if x['title']=='Zorro']

In [52]:
# [{'title': x['title'], 'Running time': x['Running time'].split('–')[0]} for x in movies_with_runtime if x['title']=='Zorro']



In [53]:
movies_with_runtime_int = [{'title': x['title'], 'Running time': int(x['Running time'].split(' min', 1)[0].split('–')[0])} for x in  movies_with_runtime if type(x['Running time']) is str]
movies_with_runtime_int



[{'title': 'academy award review of walt disney cartoons', 'Running time': 41},
 {'title': 'snow white and the seven dwarfs', 'Running time': 83},
 {'title': 'pinocchio', 'Running time': 88},
 {'title': 'fantasia', 'Running time': 126},
 {'title': 'the reluctant dragon', 'Running time': 74},
 {'title': 'dumbo', 'Running time': 64},
 {'title': 'bambi', 'Running time': 70},
 {'title': 'saludos amigos', 'Running time': 42},
 {'title': 'victory through air power', 'Running time': 65},
 {'title': 'the three caballeros', 'Running time': 71},
 {'title': 'make mine music', 'Running time': 75},
 {'title': 'song of the south', 'Running time': 94},
 {'title': 'fun and fancy free', 'Running time': 73},
 {'title': 'melody time', 'Running time': 75},
 {'title': 'so dear to my heart', 'Running time': 82},
 {'title': 'the adventures of ichabod andmr. toad', 'Running time': 68},
 {'title': 'cinderella', 'Running time': 74},
 {'title': 'treasure island', 'Running time': 96},
 {'title': 'alice in wonderl

In [54]:
# How many movies have a 'Running time' value that is an integer? 430/434
len(movies_with_runtime_int)

427

In [55]:
# Store the titles of movies with 'Running time' as an int into a list named title_runtime_int.
title_runtime_int = [x['title'] for x in movies_with_runtime_int]
title_runtime_int

['academy award review of walt disney cartoons',
 'snow white and the seven dwarfs',
 'pinocchio',
 'fantasia',
 'the reluctant dragon',
 'dumbo',
 'bambi',
 'saludos amigos',
 'victory through air power',
 'the three caballeros',
 'make mine music',
 'song of the south',
 'fun and fancy free',
 'melody time',
 'so dear to my heart',
 'the adventures of ichabod andmr. toad',
 'cinderella',
 'treasure island',
 'alice in wonderland',
 'the story of robin hood andhis merrie men',
 'peter pan',
 'the sword and the rose',
 'the living desert',
 'rob roy, the highland rogue',
 '20,000 leagues under the sea',
 'davy crockett, king of the wild frontier',
 'lady and the tramp',
 'the african lion',
 'the littlest outlaw',
 'the great locomotive chase',
 'davy crockett and the river pirates',
 'secrets of life',
 'westward ho the wagons!',
 'johnny tremain',
 'perri',
 'old yeller',
 'the light in the forest',
 'white wilderness',
 'tonka',
 'sleeping beauty',
 'the shaggy dog',
 "darby o'gill 

In [56]:
# The length of title_runtime_int equals movies_with_runtime_int
len(title_runtime_int)

427

In [57]:
# How many unique titles are there?
len({x:title_runtime_int.count(x) for x in title_runtime_int})

415

The count of unique movies that have an integer running time is 415.

In [58]:
# Rename the title_runtime_int list to only store the unique titles in title_runtime_int
title_runtime_int = list({x:title_runtime_int.count(x) for x in title_runtime_int}.keys())
len(title_runtime_int)

415

## Function: running_time(dictionary)

In [59]:
# Create a function that will take in a dictionary that contains the key 'Running time'.
# If the value type of 'Running time' is a str then change the value to an integer variable 'time'.
# return the 'time' variable
def running_time(dictionary):
    if type(dictionary['Running time']) is str:
        time = int(dictionary['Running time'].split(' ', 1)[0].split('–')[0])
       
    
    elif (type(dictionary['Running time']) is list) and (sum([x.find('minutes') for x in dictionary['Running time']])>0):
        time = [int(x.split(' ', 1)[0].split('–')[0]) for x in dictionary['Running time']][0]
     
    else:
        time =  dictionary['Running time']
            
    return time

In [60]:
test = df[[index for index, x in enumerate(df) if x['title'] == 'the vanishing prairie'][0]]
[int(x.split(' min', 1)[0].split('–')[0]) for x in test['Running time']][0]

60

In [61]:
test

{'title': 'the vanishing prairie',
 'Produced by': ['ben sharpsteen', 'walt disney'],
 'Written by': ['james algar', 'winston hibler'],
 'Narrated by': 'winston hibler',
 'Music by': 'paul j. smith',
 'Cinematography': 'n. paul kenworthy',
 'Edited by': 'lloyd l. richardson',
 'Production company': 'walt disney productions',
 'Distributed by': 'buena vista distribution',
 'Release date': ['august 17, 1954 ( 1954-08-17 )'],
 'Running time': ['60 minutes (vhs version)', '71 minutes (original)'],
 'Country': 'united states',
 'Language': 'english',
 'Box office': '$1.75 million (us and canadian rentals)'}

In [62]:
# Testing running_time(dictionary) function.
# Grab the dictionary for 'Savage Sam' and store the dictionary as variable, 'test'.
test = df[[index for index, x in enumerate(df) if x['title']=='savage sam'][0]]
test

{'title': 'savage sam',
 'Directed by': 'norman tokar',
 'Produced by': ['ron miller', 'walt disney'],
 'Written by': ['fred gipson', 'william turberg'],
 'Based on': 'savage sam by fred gipson',
 'Starring': ['brian keith', 'tommy kirk', 'kevin corcoran'],
 'Music by': 'oliver wallace',
 'Cinematography': 'edward colman',
 'Edited by': 'grant k. smith',
 'Production company': 'walt disney productions',
 'Distributed by': 'buena vista distribution',
 'Release date': 'june 1, 1963',
 'Running time': '103 minutes',
 'Country': 'united states',
 'Language': 'english',
 'Box office': '$3,000,000 (us/ canada)'}

In [63]:
# Testing running_time(dictionary) function with 'test' variable.
running_time(test)

103

In [64]:
# How many dictionaries in our list contain a title that is in the list title_runtime_int?
len([x for x in df if x['title'] in title_runtime_int ])

428

Question the data - Are there repeating dictionary titles maybe for remakes of movies?

In [65]:
# update df to change the keys of dictionaries that contain 'Running time' to integers using running_time(dictionary)
# Else - if there is no 'Running time' in the dictionary return the original dictionary

df = [{**x, 'Running time': running_time(x)} if 'Running time' in x else x for x in df ]

In [66]:
# How many movies do not have running time as an integer?
len([x for x in df if 'Running time' in x and type(x['Running time']) is not int])

4

In [67]:
# Which movies do not have 'Running time' as an integer?
[{'title': x['title'], 'Running time': x['Running time']} for x in df if 'Running time' in x and type(x['Running time']) is not int]

[{'title': 'the happiest millionaire',
  'Running time': ['los angeles', 'new york city', "director's cut"]},
 {'title': 'high school musical 3:  senior year', 'Running time': ['[1]']},
 {'title': 'jonas brothers:the 3d concert experience',
  'Running time': ['[1]', '[2]']},
 {'title': 'earth', 'Running time': ['[1]']}]

Since this is 4 movies we will go directly into their wiki pages to change their running time by hand.

In [68]:
# Change running time of 'The Happiest Millionarie'
df[[index for index, x in enumerate(df) if x['title']=='the happiest millionaire'][0]]['Running time'] = 118

In [69]:
# Change running time of 'High School Musical 3:  Senior Year'
df[[index for index, x in enumerate(df) if x['title']=='high school musical 3:  senior year'][0]]['Running time'] = 111

In [70]:
# Change running time of 'Jonas Brothers: the 3D Concert Experience'
df[[index for index, x in enumerate(df) if x['title']=='jonas brothers:the 3d concert experience'][0]]['Running time'] = 76

In [71]:
# Change running time of 'Earth'
df[[index for index, x in enumerate(df) if x['title']=='earth'][0]]['Running time'] = 90

In [72]:
# Check that running time updated
[{'title': x['title'], 'Running time': x['Running time']} for x in df if x['title'] in ['earth', 'jonas brothers:the 3d concert experience', 'high school musical 3:  senior year', 'the happiest millionaire'] ]

[{'title': 'the happiest millionaire', 'Running time': 118},
 {'title': 'high school musical 3:  senior year', 'Running time': 111},
 {'title': 'jonas brothers:the 3d concert experience', 'Running time': 76},
 {'title': 'earth', 'Running time': 90}]

In [73]:
# Curious - What is the average 'Running time' value
sum([x['Running time'] for x in df if 'Running time' in x and type(x['Running time']) is int])/len([x['Running time'] for x in df if 'Running time' in x and type(x['Running time']) is int])

97.55760368663594

In [74]:
# What is the max running time?
max_time = max([x['Running time'] for x in df if 'Running time' in x and type(x['Running time']) is int])
max_time

306

???? 306 minutes seems too long for a movie. Let's find out where this may be later.

In [75]:
# What is the minimum running time?
min_time = min(x['Running time'] for x in df if 'Running time' in x)
min_time

22

This seems too short for a feature length film.
\
\
Let's look at where these minimum and maximums are.

In [76]:
# find dictionaries where running time is equal to max_time. Store the index, and title in a list.
[(index, x['title']) for index, x in enumerate(df) if ('Running time' in x) and (x['Running time']==max_time) ]

[(446, 'night at the museum')]

Going back to the wiki page for 'Night at the Museum' This is a movie that is set for future production. This movie is part of a series, and looks like the wiki pages shows running time as running time for all the movies. We can remove this dictionary from the df list.

In [77]:
# find dictionaries where running time is equal to min_time. Store the index, and title in a list.
[(index, x['title']) for index, x in enumerate(df) if ('Running time' in x) and (x['Running time']==min_time)]

[(43, 'zorro'), (48, 'zorro')]

Going back to the wiki page for 'Zorro' these are both a part of series that aired on abc. These are only episodes that are 22 minutes long. We can remove these dictionaries from the df list.


In [78]:
# Let's make sure there is only one 'Night at the Museum' movie before we remove the dictionary.
[(index, x['title']) for index, x in enumerate(df) if x['title']=='night at the museum']

[(446, 'night at the museum')]

In [79]:
# Let's make sure there are only two 'Zorro' movies before we remove the dictionaries.
[(index,x['title']) for index, x in enumerate(df) if x['title']=='zorro']

[(43, 'zorro'), (48, 'zorro')]

In [80]:
# Let's remove these movies from the list.
remove_list = [x for x in df if x['title'] in ['night at the museum', 'zorro']]

In [81]:
for movie in remove_list:
    df.remove(movie)
    
# the length should be 444 if we are dropping 3 movies.    
len(df)

444

In [82]:
# Now what is the movie with the max time?
max_time = max([x['Running time'] for x in df if 'Running time' in x and type(x['Running time']) is int])
max_time

167

In [83]:
[(index, x) for index, x in enumerate(df) if ('Running time' in x) and (x['Running time']==max_time) ]

[(302,
  {'title': "pirates of the caribbean:  at world's end",
   'Directed by': 'gore verbinski',
   'Produced by': 'jerry bruckheimer',
   'Written by': ['ted elliott', 'terry rossio'],
   'Based on': ['characters by ted elliott terry rossio stuart beattie jay wolpert',
    'ted elliott',
    'terry rossio',
    'stuart beattie',
    'jay wolpert',
    'pirates of the caribbean by walt disney'],
   'Starring': ['johnny depp',
    'orlando bloom',
    'keira knightley',
    'stellan skarsgård',
    'bill nighy',
    'chow yun-fat',
    'geoffrey rush',
    'jack davenport',
    'kevin r. mcnally',
    'jonathan pryce'],
   'Music by': 'hans zimmer',
   'Cinematography': 'dariusz wolski',
   'Edited by': ['craig wood', 'stephen rivkin'],
   'Production companies': ['walt disney pictures', 'jerry bruckheimer films'],
   'Distributed by': 'buena vista pictures',
   'Release date': ['may 19, 2007 ( 2007-05-19 ) ( disneyland resort )',
    'may 25, 2007 ( 2007-05-25 ) (united states)'],
 

In [84]:
# Now what is the movie with the min time?
min_time = min(x['Running time'] for x in df if 'Running time' in x)
min_time

40

In [85]:
[(index, x) for index, x in enumerate(df) if ('Running time' in x) and (x['Running time']==min_time)]

[(273,
  {'title': 'sacred planet',
   'Directed by': ['jon long', 'hairul salleh askor'],
   'Produced by': ['jon long', 'karen fernandez long'],
   'Written by': ['karen fernandez long', 'jon long'],
   'Starring': ['arapata mckay',
    'tsaan ciqae',
    'mae tui',
    'cy peck jr.',
    'mutang urud'],
   'Narrated by': 'robert redford',
   'Cinematography': 'william reeve',
   'Edited by': 'jon long',
   'Production company': 'walt disney pictures',
   'Distributed by': 'buena vista pictures',
   'Release date': ['april 22, 2004 ( 2004-04-22 )'],
   'Running time': 40,
   'Country': ['canada', 'malaysia', 'united states'],
   'Language': 'english',
   'Box office': '$1,108,356'}),
 (291,
  {'title': 'roving mars',
   'Directed by': 'george butler',
   'Produced by': ['frank marshall'],
   'Written by': ['robert andrus', 'george butler'],
   'Narrated by': 'paul newman (introduction only)',
   'Music by': ['philip glass', 'sigur rós'],
   'Cinematography': 't.c. christensen',
   'E

In [86]:
write_to_json(df,'movie_dictionaries_cleaned.json')

In [91]:
df = load_data('movie_dictionaries_cleaned.json')
len(df)

444

In [94]:
[(index, x['title']) for index,x in enumerate(df) if x.get('Running time', 'N\A')=='N\A']

[(324, 'lilly the witch:the dragon and the magic book'),
 (327, 'walt & el grupo'),
 (344, 'anaganaga o dheerudu'),
 (401, 'ghost of the mountains'),
 (435, 'raya and the last dragon'),
 (436, 'cruella'),
 (437, 'jungle cruise'),
 (438, 'the beatles: get back'),
 (439, 'encanto'),
 (440, 'the little mermaid'),
 (441, 'peter pan & wendy'),
 (442, 'home alone'),
 (443, 'shrunk')]

In [115]:
# Dictionary had 'Budget' data
len([x.get('Budget') for x in df if x.get('Budget') is not None])

284

In [116]:
# Dictionary has 'Box office' data
len([x.get('Box office') for x in df if x.get('Box office') is not None])

368

In [117]:
# Dictionary has both 'Box office' and 'Budget data'
len([(x.get('Budget'), x.get('Box office')) for x in df if x.get('Budget') is not None and x.get('Box office') is not None])

272

In [148]:
for x in df:
    if x.get('Budget') is not None and type(x.get('Budget')) is str:
        try:  
            budget = float(x.get('Budget').strip('$').strip('million').replace(',', ""))
            print(budget)
        
        except Exception as e:
            print(x.get('Budget'))
            continue

1.49
2.6
2.28
600000.0
950000.0
858000.0
788000.0
1.35
2.125
1.5
1.5
2.9
1800000.0
3.0
4.0
2.0
300000.0
1.8
5.0
4.0
700000.0
6.0
under $1 million or $1,250,000
2.0
2.5
4.0
3.6
3.0
3.0
3.0
$4.4–6 million
4.0
5.0
5.0
4.0
6.3
5.0
8.0
au$1 million
5.0
7.5
10.0
$3.5 to 4 million
5.25
20.0
9.0
$6-8 million
20.0
18.0
12.0
14.0
us$ 17 million
5.0
unknown
20.0
11.0
28.0
44.0
14.0
9.0
a$8.7 million
31.0
18.0
5.0
40.0
20.0
14.0
60 million norwegian kroner (around $8.7 million in 1989)
$35-40 million
25.0
15.0
40.0
14.0
28.0
12.0
6.5
28.0
17.0
30.0
13.0
45.0
31.0
22.0
30.0
22.0
55.0
24.0
15.0
12.0
30.0
31.0
38.0
100.0
15.0
75.0
32.0
85.0
55.0
3.0
16.0
80.0
30.0
24.0
90.0
15.0
30.0
120.0
90.0
65.0
5.0
130.0
90.0
10.0
90.0
$80–$85 million
$15 million –$30 million
127.5
65.0
30.0
85.0
100.0
23.0
$90–120 million
26.0
25.0
115.0
33.0
20.0
22.0
80.0
35.0
15.0
65.0
140.0
20.0
46.0
13.0
20.0
17.0
94.0
140.0
26.0
46.0
90.0
10.0
28.0
15.0
110.0
110.0
45.0
92.0
100.0
20.0
56.0
25.0
50.0
35.0
35.0
25.0
150.0
